CMP: add `-nonmatched_error_nonces` / `OSSL_CMP_OPT_NONMATCHED_ERROR_NONCES` option

DDvO opened the pull request:
Accept missing or non-matching transactionID or recipNonce values in error messages received.

This can be helpful in case the server (or other peer) cannot provide a proper error message,
for instance if was unable to parse the ASN.1 encoding of a request message.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

DDvO committed:
CMP: Optionally accept missing or non-matching transactionID or recipNonce values

This is done for error messages received if OSSL_CMP_OPT_NONMATCHED_ERROR_NONCES
is set or the respective -nonmatched_error_nonces CLI option is given.

Can be helpful when the server (or other peer) cannot provide a proper error message header,
for instance if was unable to parse the ASN.1 encoding of a request message.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

DDvO commented:
@openssl/committers can we please get this done, 
now that 4.0 is released and hopefully there is more reviewing breath?

esyr commented on CHANGES.md:
```diff
@@ -379,6 +385,12 @@ OpenSSL Releases
+   This work was sponsored by Siemens AG.
+
+   *David von Oheimb*
+
```
This hunk can be dropped by now, I guess.

DDvO replied:
This adds a missing `CHANGES.md` entry, which I forgot to include in #28015).
I just moved this to a separate commit and made its title more specific as suggested below for 
`-nonmatched_error_nonces` and adding the `Complements: 21722fe5545c ...` part you mention below.

esyr commented on CHANGES.md:
```diff
@@ -76,6 +76,12 @@ OpenSSL Releases
 
    *Daniel Kubec*
 
+ * CMP: added `-nonmatched_error_nonces` option (also at API level).
```
It can be probably more specific, "Added `OSSL_CMP_OPT_NONMATCHED_ERROR_NONCES` option for `OSSL_CMP_CTX` and a corresponding `-nonmatched_error_nonces` option for `openssl cmp` command."

DDvO replied:
done

esyr commented on apps/cmp.c:
```diff
@@ -87,7 +87,9 @@ static char *opt_srvcert = NULL;
 static int opt_unprotected_errors = 0;
+static int opt_nonmatched_error_nonces = 0;
 static int opt_ta_in_ip_extracerts = 0;
+
```
Stray newline?

DDvO replied:
Yep, removed.

esyr commented on crypto/cmp/cmp_vfy.c:
```diff
@@ -674,24 +674,35 @@ int OSSL_CMP_validate_msg(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+        int res = 0;
+
+        if (reason == 0) /* just check; used for end_of_polling */
+            return res;
```
Is it expected that `nonmatchedErrorNonces` is ignored here?

DDvO replied:
Good question - the answer is yes, because at the end of polling, we allow for two nonce values,
such that when doing the first comparison, the final outcome of the overall nonce check is not yet complete.
I've made this more clear by changing the comment here to:
`/* at end of polling, overall check is not yet complete */`
and extending the comment in `ossl_cmp_msg_check_update()` to:
```
    /*
     * Compare received nonce with the one we sent last.
     * When we received the final response at the end of polling,
     * we allow also the nonce that we sent earlier with the original request,
     * as specified in RFC 9483 section 5.1.5.
     * /
```

esyr commented on crypto/cmp/cmp_vfy.c:
```diff
@@ -674,24 +674,35 @@ int OSSL_CMP_validate_msg(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+        int res = 0;
+
+        if (reason == 0) /* just check; used for end_of_polling */
+            return res;
```
Returning here leaks `expected_str` and `actual_str`.

DDvO replied:
Oh, fixed.

esyr commented on crypto/cmp/cmp_vfy.c:
```diff
@@ -674,24 +674,35 @@ int OSSL_CMP_validate_msg(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+                                                               : actual_str);
+        } else {
+            ossl_cmp_log1(WARN, ctx, "ignoring missing or non-matching %s of error message",
+                reason == CMP_R_TRANSACTIONID_UNMATCHED ? "transactionID" : "recipNonce");
```
Does it make sense to also provide expected/actual values here?

DDvO replied:
Possibly yes, so I added it there, too.

esyr commented on crypto/cmp/cmp_vfy.c:
```diff
@@ -674,24 +674,35 @@ int OSSL_CMP_validate_msg(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
-                                                           : actual_str);
+        char *expected_str = i2s_ASN1_OCTET_STRING(NULL, expected);
+        char *actual_str = actual == NULL ? NULL : i2s_ASN1_OCTET_STRING(NULL, actual);
+        int strict = !ctx->nonmatchedErrorNonces;
```
`const int strict = bodytype != OSSL_CMP_PKIBODY_ERROR || !ctx->nonmatchedErrorNonces;`, maybe, to have all the logic that defines `strict` in one place?

DDvO replied:
Nice idea, done.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

esyr commented on crypto/cmp/cmp_local.h:
```diff
@@ -55,6 +55,7 @@ struct ossl_cmp_ctx_st {
      * certificate responses (ip/cp/kup), revocation responses (rp), and PKIConf
      */
     int unprotectedErrors;
+    int nonmatchedErrorNonces; /* accept missing/wrong transactionID or recipNonce in error msgs */
```
This could've bin a bit field.

Also, `pahole` report for `struct ossl_cmp_ctx_st` is pretty spicy:
```
        /* size: 536, cachelines: 9, members: 75 */
        /* sum members: 504, holes: 8, sum holes: 32 */
        /* last cacheline: 24 bytes */
```
Not that anyone cares, however.

DDvO replied:
Yeh, right - in retrospect, the whole structure should have been more structured ;-)

esyr commented on CHANGES.md:
```diff
@@ -76,6 +76,12 @@ OpenSSL Releases
 
```
Commit message can be more elaborate, clarifying as to why this is needed (essentially borrowing the blurb from the PR description). Also, please try to keep the commit message subject length under 80 characters.

DDvO replied:
done

esyr commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -335,6 +335,12 @@ B<WARNING:> This setting leads to unspecified behavior and it is meant
 exclusively to allow interoperability with server implementations violating
 RFC 9810.
 
+=item B<OSSL_CMP_OPT_NONMATCHED_ERROR_NONCES>
```
I would expect that flags are mentioned in `NAME`/`SYNOPSIS`, but apparently they aren't.

DDvO replied:
I've just checked: most flags seem not mentioned in the `NAME` / `SYNOPSIS` sections.

esyr commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -888,12 +894,16 @@ OSSL_CMP_CTX_set1_serialNumber(),
 OSSL_CMP_CTX_get0_libctx(), OSSL_CMP_CTX_get0_propq(), and
 OSSL_CMP_CTX_get0_validatedSrvCert() were added in OpenSSL 3.2.
 
-OSSL_CMP_CTX_get0_geninfo_ITAVs() and
-the B<OSSL_CMP_OPT_NO_CACHE_EXTRACERTS> option were added in OpenSSL 3.3.
+The B<OSSL_CMP_OPT_NO_CACHE_EXTRACERTS> option as well as
+OSSL_CMP_CTX_get0_geninfo_ITAVs() were added in OpenSSL 3.3.
```
Not sure why this change is a part of this commit.

DDvO replied:
Right, is doc fixup should have been in separate commit, done now.

esyr commented on doc/man3/OSSL_CMP_CTX_new.pod:
```diff
@@ -888,12 +894,16 @@ OSSL_CMP_CTX_set1_serialNumber(),
 Support for central key generation, requested via B<OSSL_CRMF_POPO_NONE>,
 was added in OpenSSL 3.5.
 
+The B<OSSL_CMP_OPT_PERMIT_TA_IN_EXTRACERTS_FOR_IR> option was added in OpenSSL 4.0.
```
Same here;  ideally (but not necessarily), it should be a separate small commit with `Complements: 21722fe5545c "OSSL_CMP_OPT_PERMIT_TA_IN_EXTRACERTS_FOR_IR: use new TA(s) for whole transaction (also pkiconf); update doc"`.

DDvO replied:
Nice - I was not aware of that `Complements:` convention; used now as you suggested, 
as part of the separate commit already mentioned above.

esyr commented on test/cmp_vfy_test.c:
```diff
@@ -477,7 +479,7 @@ static void setup_check_update(CMP_VFY_TEST_FIXTURE **fixture, int expected,
     (*fixture)->allow_unprotected_cb = cb;
     (*fixture)->additional_arg = arg;
-    (*fixture)->msg = OSSL_CMP_MSG_dup(ir_rmprotection);
+    (*fixture)->msg = msg == NULL ? load_pkimsg(error_protected_f, libctx) : OSSL_CMP_MSG_dup(msg);
```
Ugh, this is a hack, why `error_protected_f` can't be loaded/freed the same way as `ir_rmprotection` in `setup_tests()`?

DDvO replied:
All right, done so.

esyr reviewed (changes requested):
Please fix `expected_str`/`actual_str` leak.

DDvO committed:
add HISTORY and CHANGES.md entry for OSSL_CMP_OPT_PERMIT_TA_IN_EXTRACERTS_FOR_IR introduced in 4.0

Complements: 21722fe5545c "OSSL_CMP_OPT_PERMIT_TA_IN_EXTRACERTS_FOR_IR: use new TA(s) for whole transaction (also pkiconf); update doc"

DDvO committed:
OSSL_CMP_CTX_new.pod: fix HISTORY entry for OSSL_CMP_CTX_get0_geninfo_ITAVs() and OSSL_CMP_OPT_NO_CACHE_EXTRACERTS

Complements 8d9c8b3ff "CMP doc: add missing text on OSSL_CMP_OPT_NO_CACHE_EXTRACERTS to OSSL_CMP_CTX_new.pod and ossl_cmp_msg_check_update.pod"

DDvO commented:
Thanks @esyr for your thoughtful review.
I've handled all comments.

DDvO commented:
@esyr did you notice that I handled all your comments?

jogme commented:
> @esyr did you notice that I handled all your comments?

He is on PTO and will get back to you. Don't worry

esyr reviewed (approved):
LGTM, thank you!

One small nit is a missing colon after "Complements" in "OSSL_CMP_CTX_new.pod: fix HISTORY entry for OSSL_CMP_CTX_get0_geninfo_ITAVs() and OSSL_CMP_OPT_NO_CACHE_EXTRACERTS".

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Merged to master - thanks @esyr and @t8m for your reviews and approvals.
